### PR TITLE
Optimize Enum.fetch/2 for negative and out of bound indexes

### PR DESCRIPTION
**Pull request by eksperimental**
This function has been highly-optimized, through thorough benchmarking.

There were serious issues when dealing with big lists and maps when an out of
bound or a negative index was given. Eg. fetching an out-of-bound index in a
1,000-element map, was reduced to 0.18% of the original item. Or same case
with a 1,000-element list time was reduced to a 10,5%.

Enumerables are no longer reversed when dealing with negative indexes.
The much cheaper `Enum.count/1` function is used now.

Benchmark .exs file are published here:
https://gist.github.com/eksperimental/dd5d4fec5988ba51f2191e5a754ed710

They have been done with lists and maps of the following configurations: 0, 1,
100, 1K, 100K, 1M elements, and indexes being 0 or positive and negative,
retrieving first, middle, last elements and and out-of-bound indexes.

There is one remark to mention about the benchmarks. In list and maps with 1K,
100K, and 1M elements, it shouldn't take into acount the porcentages in the
table when dealing with fist (negative index) and last (negative index). The
worst case scenario for negative indexes in the previous implementation was when
it was given the negative represenation of the first element in the enumerable.
Since the list was reversed, it was the last element in the reversed this. But
since the new implementation avoids using reverse, the worst case scenario from
the previous implementation (the first element with negative index) should be
compared to the worst case scenario (the last element with negative index), and
the best case scenarios should be compared as well (the last element with
negative index of previous implementation, with the first element with negative
index of the current implementation).

To faciliate this a table is provided, that shows that in every single case the
current implementation outperforms the previous one.

```
Elixir v1.3.1                                    | Elixir v1.4.0-dev

:"Enum.fetch/2: list, 1000 elements"
benchmark name      iterations   average time    | benchmark name      iterations   average time
last (neg index)       1000000   4.16 µs/op      | first (neg index)      5000000   1.59 µs/op
first (neg index)       500000   15.51 µs/op     | last (neg index)        500000   12.53 µs/op

:"Enum.fetch/2: list_big, 100000 elements"
last (neg index)         10000   349.24 µs/op    | first (neg index)        50000   146.11 µs/op
first (neg index)         5000   1428.20 µs/op   | last (neg index)          5000   1249.18 µs/op

:"Enum.fetch/2: list_huge, 1000000 elements"
last (neg index)           500   9655.58 µs/op   | first (neg index)         5000   1601.54 µs/op
first (neg index)          500   20617.46 µs/op  | last (neg index)           500   12761.61 µs/op

:"Enum.fetch/2: map, 1000 elements"
last (neg index)         50000   123.35 µs/op    | first (neg index)       200000   26.71 µs/op
first (neg index)        50000   123.22 µs/op    | last (neg index)         50000   94.47 µs/op

:"Enum.fetch/2: map_big, 100000 elements"
last (neg index)           500   12032.04 µs/op  | first (neg index)         1000   3228.43 µs/op
first (neg index)          500   12622.49 µs/op  | last (neg index)           500   10026.25 µs/op

:"Enum.fetch/2: map_huge, 1000000 elements"
last (neg index)            50   126813.58 µs/op | first (neg index)          100   32006.77 µs/op
first (neg index)           50   132636.08 µs/op | last (neg index)            50   107091.40 µs/op
```

**Decrease in performace.**
There are very few cases where performance was decreased.

__:"Enum.fetch/2: list, 1000 elements"__
```
 middle                      +3.91%
 first                       +4.00%
```

__:"Enum.fetch/2: list_big, 100000 elements"__
```
 last                        +1.25%   (1077.57 vs. 1091.07 µs/op)
 middle                      +1.37%   (543.61 vs. 551.08 µs/op)
 first                       +43.48%  (0.05 vs. 0.08 µs/op)
```
__:"Enum.fetch/2: list_empty, 0 elements"__
```
 middle                      +1.55%
```

__:"Enum.fetch/2: list_huge, 1000000 elements"__
```
 last                        +0.08%  (11518.49 vs. 11527.60 µs/op)
 first                       +30.74% (0.05 vs. 0.07 µs/op)
```

__:"Enum.fetch/2: list_single, 1 elements"__
```
 last                        +7.12%  (0.06 vs. 0.06 µs/op)
```

__:"Enum.fetch/2: map, 1000 elements"__
```
 first                       +2.47%
 middle                      +3.15%
```

When the decrease is > 5%, the absolute difference is maximum 0.03 µs/op.

### CODE REVIEW

**Comment by eksperimental** [comment]
full output can be found [here](https://gist.github.com/eksperimental/dd5d4fec5988ba51f2191e5a754ed710#file-2_enum_fetch_benchmarking_output-txt)
```
$ mix bench.cmp -d percent
bench/snapshots/2016-07-06_11-04-00.snapshot vs
bench/snapshots/2016-07-06_11-15-30.snapshot

## :"Enum.fetch/2: list, 1000 elements"
out of bound (neg index)    -89.82%
first (neg index)           -89.73%
out of bound                -89.65%
middle (neg index)          -24.77%
last                        -1.84%
middle                      +3.91%
first                       +4.00%
last (neg index)            +201.28%

## :"Enum.fetch/2: list_big, 100000 elements"
out of bound                -90.30%
out of bound (neg index)    -89.84%
first (neg index)           -89.77%
middle (neg index)          -24.19%
last                        +1.25%
middle                      +1.37%
first                       +43.48%
last (neg index)            +257.69%

## :"Enum.fetch/2: list_empty, 0 elements"
out of bound                -69.37%
last (neg index)            -69.00%
out of bound (neg index)    -68.77%
last                        -67.09%
first (neg index)           -20.96%
first                       -10.62%
middle (neg index)          -9.16%
middle                      +1.55%

## :"Enum.fetch/2: list_huge, 1000000 elements"
out of bound                -92.34%
first (neg index)           -92.23%
out of bound (neg index)    -92.14%
middle (neg index)          -53.99%
middle                      -1.90%
last                        +0.08%
first                       +30.74%
last (neg index)            +32.17%

## :"Enum.fetch/2: list_single, 1 elements"
out of bound                -65.43%
out of bound (neg index)    -64.94%
last (neg index)            -56.09%
first (neg index)           -4.39%
first                       -2.52%
middle (neg index)          -1.20%
middle                      -0.54%
last                        +7.12%

## :"Enum.fetch/2: map, 1000 elements"
out of bound (neg index)    -99.82%
out of bound                -99.81%
first (neg index)           -78.33%
middle (neg index)          -46.61%
last (neg index)            -23.41%
last                        -7.20%
first                       +2.47%
middle                      +3.15%

## :"Enum.fetch/2: map_big, 100000 elements"
out of bound (neg index)    -100.00%
out of bound                -100.00%
first (neg index)           -74.42%
middle (neg index)          -47.51%
last (neg index)            -16.67%
middle                      -4.40%
last                        -3.88%
first                       -2.72%

## :"Enum.fetch/2: map_empty, 0 elements"
out of bound (neg index)    -83.00%
out of bound                -82.07%
last (neg index)            -81.24%
last                        -80.58%
middle (neg index)          -74.43%
middle                      -73.55%
first                       -72.88%
first (neg index)           -72.43%

## :"Enum.fetch/2: map_huge, 1000000 elements"
out of bound (neg index)    -100.00%
out of bound                -100.00%
first (neg index)           -75.87%
middle (neg index)          -46.63%
last (neg index)            -15.55%
middle                      -4.78%
first                       -4.50%
last                        -0.14%

## :"Enum.fetch/2: map_single, 1 elements"
out of bound                -86.15%
out of bound (neg index)    -82.28%
last (neg index)            -81.77%
last                        -80.65%
first (neg index)           -75.05%
middle                      -74.26%
middle (neg index)          -72.99%
first                       -72.27%
```

**Comment by eksperimental** [comment]
not sure if all that information should be included in the commit msg though

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -717,30 +717,64 @@ defmodule Enum do
...
+  def fetch([], index) when is_integer(index),
+    do: :error
+  def fetch([elem], 0),
+    do: {:ok, elem}
+  def fetch([elem], -1),
+    do: {:ok, elem}
+  def fetch([_], index) when is_integer(index),
+    do: :error
+
+  def fetch(enumerable, index)
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This breaks `fetch` for struct. Try `File.stream!` for example.

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -717,30 +717,64 @@ defmodule Enum do
...
+  def fetch(enumerable, 0) when is_map(enumerable) and map_size(enumerable) == 1,
+    do: :error
+  def fetch(enumerable, -1) when is_map(enumerable) and map_size(enumerable) == 1,
+    do: :error
+  def fetch(enumerable, index)
+      when is_map(enumerable) and is_integer(index) and map_size(enumerable) == 1,
+    do: :error
+
+  def fetch(enumerable, index) when is_integer(index) and index < 0 do
+    case count(enumerable) + index do
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This has the same issue we have mentioned in another pull request. It will traverse the enumerable twice.

File: lib/elixir/lib/enum.ex (hunk 1 above)
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: By "this" I mean those clauses in general. Cases like `fetch(enumerable, 0)` below also seem to break regular maps.

File: lib/elixir/lib/enum.ex [hunk 3]
```diff
@@ -717,30 +717,64 @@ defmodule Enum do
...
   def fetch(enumerable, index) when is_integer(index) and index >= 0 do
-    res =
-      Enumerable.reduce(enumerable, {:cont, 0}, fn(entry, acc) ->
-        if acc == index do
-          {:halt, entry}
-        else
-          {:cont, acc + 1}
+    # this avoids unnecessarily traversing an enumerable that implements Enumerable.count/1
+    # given an invalid index
+    with {:ok, count} when (count - 1 - index) < 0 <- Enumerable.count(enumerable)
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: `Enumerable.count` is going to return `{:error, module}`. You should do `module.reduce` below instead of `Enumerable.reduce` to avoid another protocol lookup. Therefore I suggest using a `case` here. We didn't do this for `take_random` because we were not calling `Enumerable.reduce` directly in there.

File: lib/elixir/lib/enum.ex [hunk 4]
```diff
@@ -717,30 +717,64 @@ defmodule Enum do
   """
   @spec fetch(t, integer) :: {:ok, element} | :error
   def fetch(enumerable, index)
+  def fetch([], index) when is_integer(index),
+    do: :error
+  def fetch([elem], 0),
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Why not `fetch([elem | _], 0)`?

File: lib/elixir/lib/enum.ex [hunk 5]
```diff
@@ -717,30 +717,64 @@ defmodule Enum do
   """
   @spec fetch(t, integer) :: {:ok, element} | :error
   def fetch(enumerable, index)
+  def fetch([], index) when is_integer(index),
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Generally speaking, I don't like those clauses for lists. The reason is that, if we implement `fetch` for lists altogether, those cases will be equally as fast, and lists will be faster for *all* cases. So if we want to make lists faster, we should make all cases faster.

**Comment by josevalim** [comment]
@eksperimental thank you for the PR and the benchmarks.

I don't think we should special case lists and maps with two elements. My advice would be drop the clauses for maps and implement `fetch` for lists altogether (instead of only some cases).

The optimization for negative and out-of-bounds positive indexes are good, although we can make them slightly faster.

**Comment by ericmj** [comment]
I don't think we should spend too much effort on optimizing `Enum.fetch/2` for maps. I don't see where it would make sense to call `Enum.fetch/2` for maps given that the order is arbitrary.

File: lib/elixir/lib/enum.ex (hunk 1 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: sorry, but I cannot see how that would break regular maps. do you have an example?

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: yes, I could see that. So you are also saying the previous implementation (reversing the enumerable) also suffered from the same issue?

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: No because we reverse it only once and then used the reversed list.

File: lib/elixir/lib/enum.ex (hunk 1 above)
Comment (josevalim) [inline lib/elixir/lib/enum.ex]:
```iex
iex> Enum.fetch %{foo: 1}, 0
{:ok, {:foo, 1}}
```

The clause below will make it return `:error`.

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: But in the worst case scenario it will traverse the list twice when reversing. Let's say you give a negative index that belongs to the first element in the original list.

I couldn't work my head around this one. But i think reverse gets more expensive as the list grows bigger.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: I was not commenting on the performance, I was commenting on the correctness. :) Those functions should traverse the enumerable just once. :)

File: lib/elixir/lib/enum.ex (hunk 3 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: got it. that's great

File: lib/elixir/lib/enum.ex (hunk 1 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: you are right, 0 and -1 should have returned the only element in map, not :error
anyway... we are ditching those clauses

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: Is this a proper way?
```elixir
  def fetch(enumerable, index) when is_integer(index) and index < 0 do
    case Enumerable.count(enumerable) do
      {:ok, count} when (count + index) < 0 ->
        :error

      {:ok, count} ->
        fetch(enumerable, count + index)

      {:error, _} ->
        enumerable
        |> reverse([])
        |> fetch_list(abs(index + 1))
    end
  end
```

File: lib/elixir/lib/enum.ex (hunk 5 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: moving all of the to private func `fetch_list/2` (formerly named `do_fetch/2`)

**Comment by eksperimental** [comment]
all issues addressed, please guys tell me what you think of it.

please do not merge yet as I would like to update the commit message accordingly, and run benchmark again

**Comment by josevalim** [comment]
This one looks good to me!

File: lib/elixir/lib/enum.ex [hunk 6]
```diff
@@ -717,30 +717,61 @@ defmodule Enum do
...
-        end
-      end)
+    # this avoids unnecessarily traversing an enumerable that implements Enumerable.count/1
+    # given an invalid index
+    count_res =
+      case Enumerable.count(enumerable) do
+        {:ok, count} when (count - 1 - index) < 0 ->
+          :error
+        {:ok, _} ->
+          {:ok, Enumerable}
```
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: @josevalim I'm refactoring this code, because there's still room to speed things up. I would like to know at this point, where the answer is {:ok, _}, what is the name of the target module of this protocol. so I avoid doing Enumerable.reduce, and just call module.reduce

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: sorry my question is, is there any trick to find out this module name?

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: No, you will need to call it over again.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: No, you will have to call it as Enumerable.reduce(...) as done today.

File: lib/elixir/lib/enum.ex [hunk 7]
```diff
@@ -717,30 +717,61 @@ defmodule Enum do
...
+      {:ok, count} when (count + index) < 0 ->
+        :error
+
+      {:ok, count} ->
+        fetch(enumerable, count + index)
+
+      {:error, _} ->
+        enumerable
+        |> reverse([])
+        |> fetch_list(abs(index + 1))
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: `abs(index + 1)` –> `(-index) + 1` as we already know that the `index` is negative.

File: lib/elixir/lib/enum.ex [hunk 8]
```diff
@@ -717,30 +717,61 @@ defmodule Enum do
...
-    case res do
-      {:halted, entry} -> {:ok, entry}
-      {:done, _} -> :error
+    case count_res do
+      :error ->
+        :error
+
+      {:ok, module} ->
+        reduce_res =
+          module.reduce(enumerable, {:cont, 0}, fn(entry, acc) ->
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]:
```elixir
module.reduce(enumerable, {:cont, 0}, fn
  entry, ^index ->
    {:halt, entry}
  _entry, acc ->
    {:cont, acc + 1}
end)
```

File: lib/elixir/lib/enum.ex [hunk 9]
```diff
@@ -766,7 +797,7 @@ defmodule Enum do
   def fetch!(enumerable, index) do
     case fetch(enumerable, index) do
       {:ok, h} -> h
-      :error     -> raise Enum.OutOfBoundsError
+      :error   -> raise Enum.OutOfBoundsError
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Let's stop aligning this.

File: lib/elixir/lib/enum.ex [hunk 10]
```diff
@@ -2573,9 +2604,17 @@ defmodule Enum do
 
   ## fetch
 
-  defp do_fetch([h | _], 0), do: {:ok, h}
-  defp do_fetch([_ | t], n), do: do_fetch(t, n - 1)
-  defp do_fetch([], _),      do: :error
+  @spec fetch_list(list, non_neg_integer) :: {:ok, element} | :error
+  defp fetch_list([], _index),
+    do: :error
+  defp fetch_list([elem], 0),
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: This clause is unnecessary, it's equivalent of `defp fetch_list([h | _], 0)`.

File: lib/elixir/lib/enum.ex [hunk 11]
```diff
@@ -2573,9 +2604,17 @@ defmodule Enum do
...
+    do: {:ok, elem}
+  defp fetch_list([_], _index),
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: I think this also unnecessary.

File: lib/elixir/lib/enum.ex [hunk 12]
```diff
@@ -2573,9 +2604,17 @@ defmodule Enum do
...
+    do: :error
+  defp fetch_list([h | _], 0),
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Since we touching this place we can get rig of one-letter variable and use `head` and `tail`.

File: lib/elixir/lib/enum.ex (hunk 8 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: great!

File: lib/elixir/lib/enum.ex (hunk 7 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: I thought of this as well, but i thought it may not be accepted.
actually it should be `-1`.
code updated

**Comment by eksperimental** [comment]
ready for a final review and to be merged

File: lib/elixir/lib/enum.ex [hunk 13]
```diff
@@ -717,34 +720,75 @@ defmodule Enum do
...
   @spec fetch(t, integer) :: {:ok, element} | :error
   def fetch(enumerable, index)
-  def fetch(first..last, index) when is_integer(index),
-    do: fetch_range(first, last, index)
 
-  def fetch(enumerable, index)
-      when is_list(enumerable) and is_integer(index) and index >= 0 do
-    do_fetch(enumerable, index)
+  def fetch(enumerable, index) when is_integer(index) and index < 0 do
+    with first..last <- enumerable do
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This should be a `case`.

File: lib/elixir/lib/enum.ex [hunk 14]
```diff
@@ -717,34 +720,75 @@ defmodule Enum do
...
         end
-      end)
+    end
+  end
 
-    case res do
+  defp do_fetch(enumerable, index, module) do
+    reduce_res =
+      module.reduce(enumerable, {:cont, 0},
+        fn
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Please move the `fn` to the previous line and align `end` with `module`.

**Comment by josevalim** [comment]
@lexmag, it is all yours. :)

File: lib/elixir/test/elixir/enum_test.exs [hunk 15]
```diff
@@ -151,10 +151,17 @@ defmodule EnumTest do
 
   test "fetch/2" do
     assert Enum.fetch([2, 4, 6], 0) == {:ok, 2}
+    assert Enum.fetch([2, 4, 6], -1) == {:ok, 6}
     assert Enum.fetch([2, 4, 6], 2) == {:ok, 6}
     assert Enum.fetch([2, 4, 6], 4) == :error
     assert Enum.fetch([2, 4, 6], -2) == {:ok, 4}
     assert Enum.fetch([2, 4, 6], -4) == :error
+
+    # optimizations
```
Comment (lexmag) [inline lib/elixir/test/elixir/enum_test.exs]: These tests aren't needed as we don't do any optimization for that case.

File: lib/elixir/lib/enum.ex [hunk 16]
```diff
@@ -717,34 +720,76 @@ defmodule Enum do
...
+    fetch_list(enumerable, index)
+  end
+
+  def fetch(first..last, index) when is_integer(index) do
+    fetch_range(first, last, index)
+  end
+
+  def fetch(enumerable, index) when is_integer(index) do
+    case enumerable do
+      first..last ->
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Seems we never reach this pattern since we have range matching on L#749.

File: lib/elixir/lib/enum.ex [hunk 17]
```diff
@@ -2572,9 +2617,12 @@ defmodule Enum do
...
   ## fetch
 
-  defp do_fetch([h | _], 0), do: {:ok, h}
-  defp do_fetch([_ | t], n), do: do_fetch(t, n - 1)
-  defp do_fetch([], _),      do: :error
+  defp fetch_list([], _index),
+    do: :error
+  defp fetch_list([head | _], 0),
+    do: {:ok, head}
+  defp fetch_list([_ | tail], n),
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Maybe we should `n` –> `index`, same as we do on L#2620?

File: lib/elixir/lib/enum.ex (hunk 16 above)
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: I think it would be easier to follow the code if we move all range handling into one function clause with `if index < 0 do` check inside.

File: lib/elixir/lib/enum.ex [hunk 18]
```diff
@@ -717,34 +720,76 @@ defmodule Enum do
...
+        case count_res do
+          :error ->
+            :error
+          {:ok, module} ->
+            do_fetch(enumerable, index, module)
         end
+    end
+  end
+
+  defp do_fetch(enumerable, index, module) do
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: What if we name it `fetch_enum` to match `fetch_range` and `fetch_list` naming? :bowtie:

File: lib/elixir/test/elixir/enum_test.exs (hunk 15 above)
Comment (eksperimental) [inline lib/elixir/test/elixir/enum_test.exs]: good catch.
I would still leave them to tests lists with only one element.
Are you ok with that?
I also will add a test for an empty list

Comment (lexmag) [inline lib/elixir/test/elixir/enum_test.exs]: Having more tests doesn't hurt, but let's remove `# optimizations` annotation then.

File: lib/elixir/lib/enum.ex (hunk 16 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: sorry case shouldn't have been there.
the reason is that I don't want to patter match against ranges, before i pattern match on a list.

File: lib/elixir/lib/enum.ex (hunk 18 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: :heart:  sure. but let's name it `fetch_enumerable`

**Comment by eksperimental** [comment]
@lexmag code updated after tour comments

File: lib/elixir/lib/enum.ex (hunk 18 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: @lexmag actually i remember that why i didn't change it's name. because if I do, I felt i will have to move it at the bottom of the file where all the other helpers are. and since it's not that simple what that function is doing, i felt more comfortable leaving it where it's being called from.
Should i renamed it back `do_fetch` ?
or move it to the helpers section, or leave it as it is

Comment (lexmag) [inline lib/elixir/lib/enum.ex]: I think it's fine as it is right now.

**Comment by lexmag** [comment]
Thank you @eksperimental. 💛

**Comment by eksperimental** [comment]
:tada: thank you to everyone who participated in this!
